### PR TITLE
[IMP] stock_quant_merge: Merge quant at install + hook for inherit merge

### PR DESCRIPTION
**Pull request by pedrobaeza**
This PR improves 2 things:
- It provides an installation hook to clean up existing split quants.
- It provides a hook for inheriting the values that are written in the merged quant.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
cc @oihane @clonedagain

File: stock_quant_merge/models/stock.py [hunk 1]
```diff
@@ -19,14 +19,20 @@ def _mergeable_domain(self):
                 ('lot_id', '=', self.lot_id.id),
                 ('package_id', '=', self.package_id.id),
                 ('location_id', '=', self.location_id.id),
-                ('reservation_id', '=', False),
+                ('reservation_id', '=', self.reservation_id.id),
```
Comment (clonedagain) [inline stock_quant_merge/models/stock.py:22]: is this change related to the new features or is it a bug fix?

**Comment by clonedagain** [comment]
That looks sane, but the tests are not passing anymore. Maybe you need to update them since you use another product?

File: stock_quant_merge/models/stock.py (hunk 1 above)
Comment (pedrobaeza) [inline stock_quant_merge/models/stock.py:22]: This is to allow a merging from the init hook of any of possible quants. Imagine that you have reserved 2 quants for the same move. This one allows to merge then in only one.

**Comment by pedrobaeza** [comment]
Sorry about the test. I try only with the second test (the one that I have added), and forget the first one. I'll change it.

**Comment by clonedagain** [comment]
@pedrobaeza hehe so it does happen even to the best of us :)
No problem, just ping me when it's fixed.

**Comment by pedrobaeza** [comment]
Done

**Comment by clonedagain** [comment]
Looks fine from reading the code. I'll try to test it on our big DB (~3M quants) but I fear the install may be too long for default CPU limits. If such is the case, maybe a warning should be added to the README.

**Comment by pedrobaeza** [comment]
I have already added https://github.com/OCA/stock-logistics-warehouse/pull/119/files#diff-674e9db8ba7d8f6cd1fd1c408bc5d61bR22. Isn't enough?

**Comment by clonedagain** [comment]
I was specifically thinking of a warning in the line of "this may fail to install on big databases if the CPU limit is too low"

**Comment by pedrobaeza** [comment]
Warning added on the README

**Comment by pedrobaeza** [comment]
Can you please review this again?

File: stock_quant_merge/README.rst [hunk 2]
```diff
@@ -7,13 +7,21 @@ Quant merge
 ===========
 
 Odoo splits quants each time a reservation is done: this module makes Odoo
-merge them back if they still meet the following requirements:
+merge them back when you unserve them if they still meet the following
```
Comment (gurneyalex) [inline stock_quant_merge/README.rst:10]: s/unserve/unreserve/

**Comment by gurneyalex** [comment]
this looks quite nice. :+1: once small typo is fixed. 

I'm wondering: why should I install this? Is there a perf gain on a db with a large number of quants? Something else? If so, I think it deserves mentioning in the README. 

And if there is a huge perf gain, maybe we could make a PR on odoo master?

**Comment by pedrobaeza** [comment]
@gurneyalex, this is more or less like the Windows Disk Defragmenter, but with quants instead of sectors.

**Comment by clonedagain** [comment]
@pedrobaeza I was wondering if doing this would not introduce wrong results to the dated valuation?
In the view `stock_history` the standard joins the moves and the quants, wouldn't merging the quants make it count wrong quantities?

**Comment by pedrobaeza** [comment]
Well, I don't know, because I haven't digged in that view. It depends on what quantity field takes. Can you check it?

File: stock_quant_merge/models/stock.py (hunk 1 above)
Comment (clonedagain) [inline stock_quant_merge/models/stock.py:22]: I think if we merge quants with different "in date" values, we'll be introducing wrong results in the FIFO/LIFO removal strategy.

Comment (pedrobaeza) [inline stock_quant_merge/models/stock.py:22]: OK, we can add that condition on the mergeable domain.

**Comment by clonedagain** [comment]
My understanding is that the `stock_history` counts the quantity of the moved quants, so I think that if we merge all the possible quants we'll be counting some quantities double.

So I think we should only merge quants which are used in the same moves.

File: stock_quant_merge/models/stock.py [hunk 3]
```diff
@@ -34,7 +40,8 @@ def merge_stock_quants(self):
                 for quant in quants:
                     if (self._get_latest_move(quant2merge) ==
                             self._get_latest_move(quant)):
-                        quant2merge.sudo().qty += quant.qty
+                        quant2merge.sudo().write(
+                            self._get_merged_vals(quant2merge, quant))
                         cost += quant.cost
                         cont += 1
```
Comment (clonedagain) [inline stock_quant_merge/models/stock.py:46]: If you're going to merge all the possible quants, some will have different costs, so you should be doing a weighted average like the standard does in the dated valuation (it's changed recently: https://github.com/odoo/odoo/commit/1d7c12022dcffce4792b9eec85706e19338cfa9f)

**Comment by gurneyalex** [comment]
setting needs fixing following @clonedagain's remark

**Comment by pedrobaeza** [comment]
cc @Tecnativa

**Comment by rafaelbn** [comment]
@pedrobaeza could you please review comments of this PR? Thanks

**Comment by gurneyalex** [comment]
needs rebasing

**Comment by yajo** [comment]
Code seems alright, glitches have already been commented. Waiting for fixes and rebase to do functional review (note to self: uninstall module, unreserve quant, install module, only 1 quant should be there).

**Comment by pedrobaeza** [comment]
We have lost interest in this after handling this with the customer in another way. If someone wants to take over the work, please make another PR. I close meanwhile.
